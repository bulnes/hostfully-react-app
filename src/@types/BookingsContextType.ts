export interface BookingsContextType {
  showBookings: boolean;
  setShowBookingsVisibility: (showBookings: boolean) => void;
}
