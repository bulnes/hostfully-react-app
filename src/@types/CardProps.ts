export interface CardProps {
  id: number;
  title: string;
  description: string;
  pricePerNight: number;
  maxGuests: number;
  bedrooms: number;
  bathrooms: number;
  image: string;
}
