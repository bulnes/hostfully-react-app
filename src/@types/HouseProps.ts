export interface HouseProps {
  id: number;
  title: string;
  description: string;
  pricePerNight: number;
  maxGuests: number;
  bedrooms: number;
  bathrooms: number;
  image: string;
}
