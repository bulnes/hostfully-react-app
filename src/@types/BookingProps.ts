import { HouseProps } from "./HouseProps";

export interface BookingProps extends HouseProps {
  checkInDate?: string;
  checkOutDate?: string;
}