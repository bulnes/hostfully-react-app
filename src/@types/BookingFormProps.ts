import { BookingProps } from "./BookingProps";

export interface BookingFormProps extends BookingProps {
  formType?: "booking" | "update";
}
