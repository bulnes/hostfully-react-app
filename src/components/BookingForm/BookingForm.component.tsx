import { useContext, useRef } from "react";
import { toast } from "react-toastify";
import swal from "sweetalert";
import { BookingsContext } from "../../@contexts/BookingsContext";
import { BookingProps } from "../../@types/BookingProps";
import { Button } from "../Button/Button.component";
import { isValidInfo } from "./helpers/is-valid-info";

interface BookingForm extends BookingProps {
  formType?: "booking" | "update";
}

export function BookingForm({
  id,
  title,
  description,
  pricePerNight,
  maxGuests,
  bedrooms,
  bathrooms,
  image,
  checkInDate,
  checkOutDate,
  bookingId,
  formType = "booking",
}: BookingForm) {
  const { addBooking, removeBooking, updateBooking, bookings } =
    useContext(BookingsContext);

  const checkinRef = useRef<HTMLInputElement>(null);
  const checkoutRef = useRef<HTMLInputElement>(null);

  function handleBookNow(e: React.FormEvent<HTMLFormElement>) {
    e.preventDefault();

    const bookingId = bookings.length + 1;
    const checkin = checkinRef.current?.value || "";
    const checkout = checkoutRef.current?.value || "";

    const isValid = isValidInfo({ checkin, checkout, bookings });

    if (isValid) {
      addBooking({
        id,
        title,
        description,
        pricePerNight,
        maxGuests,
        bedrooms,
        bathrooms,
        image,
        bookingId,
        checkInDate: checkin,
        checkOutDate: checkout,
      });

      toast("Your booking has been added successfully");
    }
  }

  function handleUpdateBooking(
    bookingId: number,
    checkInDate: string,
    checkOutDate: string,
    newCheckInDate: string,
    newCheckOutDate: string
  ) {
    const isValid = isValidInfo({
      checkin: checkInDate,
      checkout: checkOutDate,
      bookings,
      id: bookingId,
      oldCheckInDate: newCheckInDate,
      oldCheckOutDate: newCheckOutDate,
    });

    if (isValid) {
      const updatedBooking = {
        ...bookings.find((booking) => booking.id === bookingId),
        checkInDate: newCheckInDate,
        checkOutDate: newCheckOutDate,
      } as BookingProps;

      updateBooking(bookingId, updatedBooking);

      toast("Your booking has been updated successfully");
    }
  }

  function handleRemoveBooking(bookingId: number) {
    swal({
      title: "Are you sure?",
      text: "Once deleted, you will not be able to recover this booking!",
      icon: "warning",
      buttons: ["Cancel", "Delete"],
      dangerMode: true,
    }).then((willDelete) => {
      if (willDelete) {
        removeBooking(bookingId);

        toast("Poof! Your booking has been deleted!");
      }
    });
  }

  return (
    <form onSubmit={handleBookNow}>
      <div className="flex items-center gap-3 mt-3 mb-5">
        <div className="flex items-start flex-col gap-1">
          <label htmlFor="checkin" className="text-gray-600">
            Check-in
          </label>

          <input
            type="date"
            id="checkin"
            className="border rounded-md p-1"
            ref={checkinRef}
            defaultValue={checkInDate}
          />
        </div>

        <div className="flex items-start flex-col gap-1">
          <label htmlFor="checkout" className="text-gray-600">
            Check-out
          </label>

          <input
            type="date"
            id="checkout"
            className="border rounded-md p-1 w-full"
            ref={checkoutRef}
            defaultValue={checkOutDate}
          />
        </div>
      </div>

      {formType === "booking" && <Button type="submit">Book now</Button>}

      {formType === "update" && (
        <div className="flex items-center gap-3 mt-4">
          <Button
            type="button"
            buttonType="secondary"
            onClick={() =>
              handleUpdateBooking(
                bookingId!,
                checkInDate || "",
                checkOutDate || "",
                checkinRef.current?.value || "",
                checkoutRef.current?.value || ""
              )
            }
          >
            Update
          </Button>

          <Button
            type="button"
            buttonType="danger"
            onClick={() => handleRemoveBooking(bookingId!)}
          >
            Delete
          </Button>
        </div>
      )}
    </form>
  );
}
