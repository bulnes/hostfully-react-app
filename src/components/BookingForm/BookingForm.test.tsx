import { render } from "@testing-library/react";
import { BookingFormProps } from "../../@types/BookingFormProps";
import { BookingForm } from "./BookingForm.component";

const props = {
  title: "Title",
  description: "Description",
  pricePerNight: 100,
  maxGuests: 4,
  bedrooms: 2,
  bathrooms: 2,
  image: "https://via.placeholder.com/478x319",
} as BookingFormProps;

describe("Component: BookingForm", () => {
  it("should be defined", () => {
    expect(BookingForm).toBeDefined();
  });

  it("should render correctly", () => {
    const wrapper = render(<BookingForm {...props} />);
    expect(wrapper).toBeDefined();
  });
});
