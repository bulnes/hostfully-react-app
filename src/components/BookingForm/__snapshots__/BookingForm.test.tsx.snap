// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: BookingForm should be represented by the snapshot 1`] = `
<div>
  <form>
    <div
      class="flex items-center gap-3 mt-3 mb-5"
    >
      <div
        class="flex items-start flex-col gap-1"
      >
        <label
          class="text-gray-600"
          for="checkin"
        >
          Check-in
        </label>
        <input
          class="border rounded-md p-1"
          id="checkin"
          type="date"
          value=""
        />
      </div>
      <div
        class="flex items-start flex-col gap-1"
      >
        <label
          class="text-gray-600"
          for="checkout"
        >
          Check-out
        </label>
        <input
          class="border rounded-md p-1 w-full"
          id="checkout"
          type="date"
          value=""
        />
      </div>
    </div>
    <button
      class="bg-blue-500 hover:bg-blue-700 text-white border border-blue-500 font-bold py-2 px-4 rounded"
      type="submit"
    >
      Book now
    </button>
  </form>
</div>
`;
